File: modules/core/src/model/PurposeRestriction.ts
import {Cloneable} from '../Cloneable';
import {TCModelError} from '../errors';
import {RestrictionType} from './RestrictionType';

export class PurposeRestriction extends Cloneable<PurposeRestriction> {

  public static hashSeparator = '-';

  private purposeId_: number;
  public restrictionType: RestrictionType;

  /**
   * constructor
   *
   * @param {number} purposeId? - may optionally pass the purposeId into the
   * constructor
   * @param {RestrictionType} restrictionType? - may
   * optionally pass the restrictionType into the constructor
   * @return {undefined}
   */
  public constructor(purposeId?: number, restrictionType?: RestrictionType) {

    super();

    if (purposeId !== undefined) {

      this.purposeId = purposeId;

    }

    if (restrictionType !== undefined) {

      this.restrictionType = restrictionType;

    }

  }

  public static unHash(hash: string): PurposeRestriction {

    const splitUp: string[] = hash.split(this.hashSeparator);
    const purpRestriction: PurposeRestriction = new PurposeRestriction();

    if (splitUp.length !== 2) {

      throw new TCModelError('hash', hash);

    }

    purpRestriction.purposeId = parseInt(splitUp[0], 10);
    purpRestriction.restrictionType = parseInt(splitUp[1], 10);

    return purpRestriction;

  }
  public get hash(): string {

    if (!this.isValid()) {

      throw new Error('cannot hash invalid PurposeRestriction');

    }

    return `${this.purposeId}${PurposeRestriction.hashSeparator}${this.restrictionType}`;

  }

  /**
   * @return {number} The purpose Id associated with a publisher
   * purpose-by-vendor restriction that resulted in a different consent or LI
   * status than the consent or LI purposes allowed lists.
   */
  public get purposeId(): number {

    return this.purposeId_;

  }
  /**
   * @param {number} idNum - The purpose Id associated with a publisher
   * purpose-by-vendor restriction that resulted in a different consent or LI
   * status than the consent or LI purposes allowed lists.
   */
  public set purposeId(idNum: number) {

    this.purposeId_ = idNum;

  }

  public isValid(): boolean {

    return (
      Number.isInteger(this.purposeId) &&
      this.purposeId > 0 &&
      (
        this.restrictionType === RestrictionType.NOT_ALLOWED ||
        this.restrictionType === RestrictionType.REQUIRE_CONSENT ||
        this.restrictionType === RestrictionType.REQUIRE_LI
      )
    );

  }
  public isSameAs(otherPR: PurposeRestriction): boolean {

    return (this.purposeId === otherPR.purposeId &&
      this.restrictionType === otherPR.restrictionType);

  }

}

File: GDPR-Phase2-Java/src/main/java/com/gdpr/phase2/model/PurposeRestriction.java
package com.gdpr.phase2.model;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;


public class PurposeRestriction {
    public static String hashSeparator = "-";
    private int purposeId = 0;
    public RestrictionType restrictionType;

    public PurposeRestriction(int purposeId, RestrictionType restrictionType) {
        this.purposeId = purposeId;
        this.restrictionType = restrictionType;
    }

    public PurposeRestriction() {
    }

    public static PurposeRestriction unHash(String hash) {
        List<String> splitUp = Arrays.asList(hash.split(hashSeparator));
        PurposeRestriction purpRestriction = new PurposeRestriction();
        if(splitUp.size()!=2) {
            // throw error
        }
        purpRestriction.purposeId = Integer.parseInt(splitUp.get(0));
        purpRestriction.restrictionType.setType(Integer.parseInt(splitUp.get(1)));
        return purpRestriction;

    }

    public String getHash() {
        if(!isValid()) {
            // throw error
        }
        return purposeId + PurposeRestriction.hashSeparator + restrictionType.getType();
    }

    private boolean isValid() {
        return purposeId!=0 && restrictionType!= null;
    }

    public int getPurposeId() {
        return purposeId;
    }

    public void setPurposeId(int purposeId) {
        this.purposeId = purposeId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof PurposeRestriction)) return false;
        PurposeRestriction that = (PurposeRestriction) o;
        return purposeId == that.purposeId &&
                restrictionType == that.restrictionType;
    }

    @Override
    public int hashCode() {
        return Objects.hash(purposeId, restrictionType);
    }
}
